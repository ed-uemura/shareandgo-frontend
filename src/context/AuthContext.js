import React, { createContext, useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { auth, db } from '../firebaseConfig';
import {
	createUserWithEmailAndPassword,
	signInWithEmailAndPassword,
	signOut,
	deleteUser,
} from 'firebase/auth';
import {
	doc,
	collection,
	addDoc,
	setDoc,
	getDoc,
	deleteDoc,
	updateDoc,
	query,
	where,
	getDocs,
} from 'firebase/firestore';

const Context = createContext();

const AuthContext = ({ children }) => {
	const [authenticated, setAuthenticated] = useState(false);
	const [loading, setLoading] = useState(true);
	const [userData, setUserData] = useState({});
	const [loggedUser, setLoggedUser] = useState({});
	const [destination, setDestination] = useState({});
	const [setupRide, setSetupRide] = useState({});
	const [rideData, setRideData] = useState([]);
	const [rideSelection, setRideSelection] = useState({});
	const navigate = useNavigate();

	useEffect(() => {
		if (!loggedUser) {
			setAuthenticated(false);
		} else if (loggedUser) {
			setAuthenticated(true);
		}

		if (authenticated) {
			const getUser = async () => {
				try {
					const docRef = doc(db, 'users', loggedUser.uid);
					const docSnap = await getDoc(docRef);

					if (docSnap.exists()) {
						setUserData(docSnap.data());
					} else {
						console.log('No document');
					}
				} catch (err) {
					console.log(err.message);
				}
			};

			getUser();
		}

		setLoading(false);
	}, [loggedUser, setUserData, authenticated]);

	//Handle Registration
	const handleRegistration = async (values) => {
		try {
			let userUid;
			await createUserWithEmailAndPassword(
				auth,
				values.email,
				values.password
			).then((cred) => {
				userUid = cred.user.uid;
			});

			if (values.accountType === 'Passenger') {
				await setDoc(doc(db, 'users', userUid), {
					firstName: values.firstName,
					lastName: values.lastName,
					accountType: values.accountType,
					phone: values.phone,
					address: {
						street: values.address,
						city: values.city,
						province: values.province,
						zip: values.zip,
					},
					photo: values.photo,
					payment: {
						nameOnCard: values.nameCard,
						cardNumber: values.cardNumber,
						cardExpirationDate: values.expirationDate,
						cardCsc: values.csc,
						cardZipCode: values.cardZipCode,
					},
					rides: [],
					evaluation: '5',
				})
					.then(() => {
						return navigate('/main-passenger');
					})
					.catch((err) => {
						console.log(err.message);
						handleSignOut();
						return navigate('/');
					});
			} else if (values.accountType === 'Driver') {
				await setDoc(doc(db, 'users', userUid), {
					firstName: values.firstName,
					lastName: values.lastName,
					accountType: values.accountType,
					phone: values.phone,
					address: {
						street: values.address,
						city: values.city,
						province: values.province,
						zip: values.zip,
					},
					photo: values.photo,
					payment: {
						bankName: values.bankName,
						institutionNumber: values.institutionNumber,
						transitNumber: values.transitNumber,
						accountNumber: values.accountNumber,
					},
					rides: [],
					evaluation: '5',
				})
					.then(() => {
						return navigate('/main-driver');
					})
					.catch((err) => {
						console.log(err.message);
						handleSignOut();
						return navigate('/');
					});
			}
		} catch (err) {
			console.log(err.message);
		}
	};

	//Handle Login
	const handleSignIn = async (values) => {
		try {
			let userUid;
			await signInWithEmailAndPassword(auth, values.email, values.password)
				.then((resp) => {
					userUid = resp.user.uid;
				})
				.catch((err) => {
					console.log(err.message);
					handleSignOut();
					return navigate('/');
				});

			const docRef = doc(db, 'users', userUid);
			const docSnap = await getDoc(docRef);

			if (docSnap.exists()) {
				const docData = docSnap.data();

				if (docData.accountType === 'Driver') {
					return navigate('/main-driver');
				} else if (docData.accountType === 'Passenger') {
					return navigate('/main-passenger');
				} else {
					handleSignOut();
					return navigate('/');
				}
			} else {
				console.log('no data found');
				handleSignOut();
				return navigate('/');
			}
		} catch (err) {
			console.log(err.message);
			handleSignOut();
			return navigate('/');
		}
	};

	//Handle Logout
	const handleSignOut = async () => {
		await signOut(auth);
		return navigate('/');
	};

	//Handle Delete User
	const handleDeleteUser = async (event) => {
		event.preventDefault();
		try {
			await deleteDoc(doc(db, 'users', loggedUser.uid));
			deleteUser(auth.currentUser)
				.then(() => {
					handleSignOut();
					navigate('/');
				})
				.catch((err) => {
					console.log(err.message);
				});
		} catch (err) {
			console.log(err.message);
		}
	};

	//Handle Setup Ride
	const handlesetUpRide = async (e, values) => {
		e.preventDefault();

		const rideSetup = {
			currentLocation: values.currentLocation,
			destinationLocation: values.destinationLocation,
			rideDate: values.rideDate,
			rideTime: values.rideTime,
			numOfSeats: values.numOfSeats,
			rideDetails: values.rideDetails,
			driver: {
				uid: auth.currentUser.uid,
				photo: userData.photo,
				firstName: userData.firstName,
				lastName: userData.lastName,
				rating: userData.evaluation,
			},
			passengers: [],
			passengerNum: 0,
			price: '82.40',
			reviews: [],
		};

		setSetupRide(rideSetup);

		navigate('/create-ride');
	};

	//Handle create ride
	const handleCreateRide = async (e, values) => {
		e.preventDefault();

		let rideId;

		try {
			await addDoc(collection(db, 'rides'), values)
				.then((resp) => {
					rideId = resp.id;
				})
				.catch((err) => {
					console.log(err.message);
				});

			const docRef = doc(db, 'rides', rideId);
			await updateDoc(docRef, {
				uid: rideId,
			})
				.then(() => {
					return navigate(`/ride-info/${rideId}`);
				})
				.catch((err) => {
					console.log(err.message);
				});
		} catch (err) {
			console.log(err.message);
		}
	};

	//Handle Cancel Ride
	const handleDestroyRide = async (id) => {
		const docRef = doc(db, 'rides', id);

		await deleteDoc(docRef)
			.then(() => {
				navigate('/main-driver');
			})
			.catch((err) => {
				console.log(err.message);
			});
	};

	//Handle Cancel Ride
	const handleFinishRide = async (id) => {
		const docRef = doc(db, 'rides', id);

		await deleteDoc(docRef)
			.then(() => {
				navigate('/ride-complete');
			})
			.catch((err) => {
				console.log(err.message);
			});
	};

	//Handle Get Rides
	const handleGetRides = async (values) => {
		const rideDataArray = [];

		const q = query(
			collection(db, 'rides'),
			where('destinationLocation', '==', values.destinationLocation)
		);

		try {
			const querySnapshot = await getDocs(q);
			querySnapshot.forEach((doc) => {
				rideDataArray.push(doc.data());
			});
		} catch (err) {
			console.log(err.message);
		}

		setRideData(rideDataArray);
		setRideSelection(values);

		navigate('/available-drivers');
	};

	// Handle Request Ride
	const handleRequestRide = async (rideId, numOfPassengers) => {
		try {
			const docRef = doc(db, 'rides', rideId);

			await updateDoc(docRef, {
				passengerNum: numOfPassengers + 1,
				passengers: [
					{
						uid: loggedUser.uid,
						firstName: userData.firstName,
						lastName: userData.lastName,
						fullName: userData.firstName + ' ' + userData.lastName,
						photo: userData.photo,
					},
				],
			})
				.then(() => {
					return navigate('/waiting-ride');
				})
				.catch((err) => {
					console.log(err.message);
				});
		} catch (err) {
			console.log(err.message);
		}
	};

	return (
		<Context.Provider
			value={{
				authenticated,
				setAuthenticated,
				loading,
				setLoading,
				userData,
				setUserData,
				loggedUser,
				setLoggedUser,
				destination,
				setDestination,
				setupRide,
				setSetupRide,
				rideData,
				setRideData,
				rideSelection,
				setRideSelection,
				handleRegistration,
				handleSignIn,
				handleSignOut,
				handleDeleteUser,
				handlesetUpRide,
				handleCreateRide,
				handleDestroyRide,
				handleFinishRide,
				handleGetRides,
				handleRequestRide,
			}}>
			{children}
		</Context.Provider>
	);
};

export { Context, AuthContext };
