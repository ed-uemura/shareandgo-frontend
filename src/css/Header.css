.filled {
	font-variation-settings: 'FILL' 1, 'wght' 400, 'GRAD' 0, 'opsz' 48;
}

.icon {
	font-size: 30px;
}

.navbar-toggler {
	padding: 0;
}

nav.bg-dark {
	background-color: var(--SGGreen) !important;
}

.navbar-toggler .navbar-toggler-icon {
	background-image: url("data:image/svg+xml;charset=utf8,%3Csvg viewBox='0 0 32 32' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath stroke='rgba(255,255,255, 1)' stroke-width='3' stroke-linecap='round' stroke-miterlimit='10' d='M4 8h24M4 16h24M4 24h24'/%3E%3C/svg%3E");
}

.navbar-toggler.navbar-toggler {
	border-color: var(--SGAlpha);
}

.navbar-nav {
	--bs-nav-link-color: white;
}

.btn-close {
	background: transparent
		url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16' fill='%23000'%3e%3cpath d='M.293.293a1 1 0 0 1 1.414 0L8 6.586 14.293.293a1 1 0 1 1 1.414 1.414L9.414 8l6.293 6.293a1 1 0 0 1-1.414 1.414L8 9.414l-6.293 6.293a1 1 0 0 1-1.414-1.414L6.586 8 .293 1.707a1 1 0 0 1 0-1.414z'/%3e%3c/svg%3e")
		center/1em auto no-repeat;
	color: white;
	opacity: 1;
}

.offcanvas.offcanvas-top {
	height: 100%;
}

.offcanvas-body,
.offcanvas-header {
	background-color: var(--SGGreen);
	color: white;
	font-size: var(--SGTextAssetSize);
	font-weight: 700;
}

.offcanvas-body li {
	font-size: var(--SGTitleSize);
}
