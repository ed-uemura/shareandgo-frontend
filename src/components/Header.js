import React from 'react';
import { Link } from 'react-router-dom';
import '../css/Header.css';
import ButtonSecondary from './ButtonSecondary';
import LogoHeader from '../images/ShareGoLogo-header.svg';

const Header = () => {
	return (
		<>
			<nav className='navbar navbar-dark bg-dark mb-5'>
				<div className='container d-flex align-items-center'>
					<button
						className='navbar-toggler'
						type='button'
						data-bs-toggle='offcanvas'
						data-bs-target='#offcanvasDarkNavbar'
						aria-controls='offcanvasDarkNavbar'>
						<span className='material-symbols-outlined icon'>menu</span>
					</button>

					<Link
						to={'/main-passenger'}
						className='nav-brand flex-fill text-center'>
						<img src={LogoHeader} alt='Share & Go Logo' />
					</Link>

					<Link to={'/messages'}>
						<span className='material-symbols-outlined icon filled'>
							notifications
						</span>
					</Link>

					<div
						className='offcanvas offcanvas-top text-bg-dark'
						tabIndex='-1'
						id='offcanvasDarkNavbar'
						aria-labelledby='offcanvasDarkNavbarLabel'>
						<div className='offcanvas-header'>
							<h5
								className='offcanvas-title mx-auto d-block'
								id='offcanvasDarkNavbarLabel'>
								<img src={LogoHeader} alt='Share & Go' />
							</h5>
							<button
								type='button'
								className='btn-close btn-close-white'
								data-bs-dismiss='offcanvas'
								aria-label='Close'></button>
						</div>
						<div className='offcanvas-body'>
							<ul className='navbar-nav flex-grow-1 pe-3 position-absolute top-50 start-50 translate-middle'>
								<li
									className='nav-item text-center'
									data-bs-dismiss='offcanvas'>
									<Link to={'/user-account'} className='nav-link'>
										My Account
									</Link>
								</li>

								<li
									className='nav-item text-center'
									data-bs-dismiss='offcanvas'>
									<Link to={'/messages'} className='nav-link'>
										Messages
									</Link>
								</li>
							</ul>
							<div className='row' data-bs-dismiss='offcanvas'>
								<ButtonSecondary
									text='Log out'
									className='col-6 col-lg-2 col-md-3 col-sm-3 col-xs-3 btn-white position-absolute bottom-0 start-50 translate-middle-x mb-5'
									link='/'
									clickAction=''
								/>
							</div>
						</div>
					</div>
				</div>
			</nav>
		</>
	);
};

export default Header;
