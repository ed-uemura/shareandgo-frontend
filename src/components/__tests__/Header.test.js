import renderer from 'react-test-renderer'
import Header from '../Header';
import { MemoryRouter } from 'react-router-dom';
import {render, screen, cleanup} from '@testing-library/react'
import * as AuthContext from '../../context/AuthContext'


jest.mock("axios", ()=>({
    // default:{
    //     get: () =>({

    //     })
    // }
}))

afterEach(() =>{
    cleanup();
});

test('matches snapshot', () =>{ 

    const tree = renderer.create(
        <AuthContext.Context.Provider value = {{authenticated : false, loading : true, userData : {}}}>
            <MemoryRouter>
                <Header/>
            </MemoryRouter>
        </AuthContext.Context.Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});