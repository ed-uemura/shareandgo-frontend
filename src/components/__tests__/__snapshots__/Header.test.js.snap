// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<nav
  className="navbar navbar-dark bg-dark mb-5"
>
  <div
    className="container d-flex align-items-center"
  >
    <a
      className="nav-brand flex-fill text-center"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="Share & Go Logo"
        src="ShareGoLogo-header.svg"
      />
    </a>
    <div
      aria-labelledby="offcanvasDarkNavbarLabel"
      className="offcanvas offcanvas-top text-bg-dark"
      id="offcanvasDarkNavbar"
      tabIndex="-1"
    >
      <div
        className="offcanvas-header"
      >
        <h5
          className="offcanvas-title mx-auto d-block"
          id="offcanvasDarkNavbarLabel"
        >
          <img
            alt="Share & Go"
            src="ShareGoLogo-header.svg"
          />
        </h5>
        <button
          aria-label="Close"
          className="btn-close btn-close-white"
          data-bs-dismiss="offcanvas"
          type="button"
        />
      </div>
      <div
        className="offcanvas-body"
      >
        <ul
          className="navbar-nav flex-grow-1 pe-3 position-absolute top-50 start-50 translate-middle"
        >
          <li
            className="nav-item text-center"
            data-bs-dismiss="offcanvas"
          >
            <a
              className="nav-link"
              href="/user-account"
              onClick={[Function]}
            >
              My Account
            </a>
          </li>
          <li
            className="nav-item text-center"
            data-bs-dismiss="offcanvas"
          >
            <a
              className="nav-link"
              href="/messages"
              onClick={[Function]}
            >
              Messages
            </a>
          </li>
        </ul>
        <div
          className="row"
          data-bs-dismiss="offcanvas"
        >
          <a
            className="btn btn-secondary col-6 col-lg-2 col-md-3 col-sm-3 col-xs-3 btn-white position-absolute bottom-0 start-50 translate-middle-x mb-5"
            href="/"
            onClick={[Function]}
          >
            Log out
          </a>
        </div>
      </div>
    </div>
  </div>
</nav>
`;
